sphinx/ext/intersphinx.py

```diff
@@ -466,6 +470,145 @@ def missing_reference(app: Sphinx, env: BuildEnvironment, node: pending_xref,
     return resolve_reference_detect_inventory(env, node, contnode)
 
 
+class IntersphinxDispatcher(CustomReSTDispatcher):
+    """Custom dispatcher for external role.
+
+    This enables :external:***:/:external+***: roles on parsing reST document.
+    """
+
+    def role(self, role_name: str, language_module: ModuleType, lineno: int, reporter: Reporter
+             ) -> Tuple[RoleFunction, List[system_message]]:
+        if len(role_name) > 9 and role_name.startswith('external') and role_name[8] in ':+':
+            return IntersphinxRole(role_name), []
+        else:
+            return super().role(role_name, language_module, lineno, reporter)
+
+
+class IntersphinxRole(SphinxRole):
+    def __init__(self, orig_name: str) -> None:
+        self.orig_name = orig_name
+
+    def run(self) -> Tuple[List[Node], List[system_message]]:
+        assert self.name == self.orig_name.lower()
+        inventory, name_suffix = self.get_inventory_and_name_suffix(self.orig_name)
+        if inventory and not inventory_exists(self.env, inventory):
+            logger.warning(__('inventory for external cross-reference not found: %s'),
+                           inventory, location=(self.env.docname, self.lineno))
+            return [], []
+
+        role_name = self.get_role_name(name_suffix)
+        if role_name is None:
+            logger.warning(__('role for external cross-reference not found: %s'), name_suffix,
+                           location=(self.env.docname, self.lineno))
+            return [], []
+
+        result, messages = self.invoke_role(role_name)
+        for node in result:
+            if isinstance(node, pending_xref):
+                node['intersphinx'] = True
+                node['inventory'] = inventory
+
+        return result, messages
+
+    def get_inventory_and_name_suffix(self, name: str) -> Tuple[Optional[str], str]:
+        assert name.startswith('external'), name
+        assert name[8] in ':+', name
+        typ = name[8]
+        name = name[9:]
+        if typ == '+':
+            # we have an explicit inventory name, i.e,
+            # :external+inv:role:        or
+            # :external+inv:domain:role:
+            inv, name = name.split(':', 1)
+            return inv, name
+        else:
+            assert typ == ':'
+            # we look in all inventories, i.e.,
+            # :external:role:        or
+            # :external:domain:role:
+            return None, name
```

Comment:
It's better to use regexp to parse the role name, I think.
```
pattern = re.compile("external(\+\s+)?:.*")
return pattern.match(name).groups()  # => (invname, rolename)
```

Reply:
Hmm, it would make the code a bit more elegant. I'll try to take a closer look before the release, but otherwise feel free to merge and we can refactor it afterwards.

Reply:
Done.